File: src/main/java/ludwig/workspace/Workspace.java
package ludwig.workspace;

import ludwig.changes.*;
import ludwig.interpreter.Builtins;
import ludwig.model.*;
import ludwig.runtime.StdLib;
import ludwig.script.Parser;

import java.io.InputStreamReader;
import java.io.Reader;
import java.util.*;
import java.util.function.Consumer;

public class Workspace {
    public static final int MAX_PROBLEMS = 10;

    private final Map<String, Node> nodes = new HashMap<>();
    private final List<Change<?>> appliedChanges = new ArrayList<>();
    private final List<ProjectNode> projects = new ArrayList<>();
    private final List<Consumer<Change<?>>> changeListeners = new ArrayList<>();
    private ProjectNode builtins;
    private boolean batchUpdate;
    private boolean loading;

    public void init() {
        builtins = new ProjectNode().name("Runtime").id("Runtime").readonly(true);
        builtins.add(Builtins.of(StdLib.class));

        addNode(builtins);

        try {
            try (Reader reader = new InputStreamReader(Workspace.class.getResourceAsStream("/system.lw"))) {
                Parser.Companion.parse(reader, this, builtins);
            }
            try (Reader reader = new InputStreamReader(Workspace.class.getResourceAsStream("/system-tests.lw"))) {
                Parser.Companion.parse(reader, this, builtins);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public final List<Consumer<Change<?>>> changeListeners() {
        return changeListeners;
    }

    private final ChangeVisitor<Problem> changeVisitor = new ChangeVisitor<Problem>() {

        @Override
        public Problem visitInsertNode(InsertNode insert) {
            return place(insert.node(), insert);
        }

        @Override
        public Problem visitInsertReference(InsertReference insert) {
            Node ref = new ReferenceNode(node(insert.ref())).id(insert.id());
            return place(ref, insert);
        }

        @Override
        public Problem visitDelete(Delete delete) {
            Node node = node(delete.id());
            node.parent().children().remove(node);
            node.delete();
            return null;
        }

        @Override
        public Problem visitComment(Comment comment) {
            Node node = node(comment.nodeId());
            node.comment(comment.comment());
            return null;
        }

        @Override
        public Problem visitRename(Rename rename) {
            NamedNode node = node(rename.getNodeId());
            node.name(rename.name());
            return null;
        }
    };

    private Problem place(Node node, Insert insert) {
        addNode(node);
        Node parent = node(insert.parent());
        node.parent(parent);

        if (parent != null) {
            Node prev = node(insert.prev());
            Node next = node(insert.next());

            if (!parent.isOrdered()) {
                parent.add(node);
            } else {
                List items = parent.children();

                if (next == null) {
                    if (!items.isEmpty() && items.get(items.size() - 1) == prev || items.isEmpty() && prev == null) {
                        parent.add(node);
                    }
                } else if (prev == null) {
                    if (!items.isEmpty() && items.get(0) == next) {
                        items.add(0, node);
                    }
                } else {
                    int prevIndex = items.indexOf(prev);
                    int nextIndex = items.indexOf(next);

                    if (nextIndex == prevIndex + 1) {
                        items.add(nextIndex, next);
                    }
                }
            }
        }
        return null;
    }

    public List<ProjectNode> getProjects() {
        return projects;
    }

    public List<Problem> apply(List<Change<?>> changes) {
        List<Problem> problems = new ArrayList<>();
        for (int i = 0; i < changes.size(); i++) {
            Change<?> change = changes.get(i);
            batchUpdate = i < changes.size() - 1;
            Problem problem = change.accept(changeVisitor);
            if (problem != null) {
                problems.add(problem);
                if (problems.size() == MAX_PROBLEMS) {
                    break;
                }
            } else {
                changeListeners.forEach(listener -> listener.accept(change));
            }
        }

        if (problems.isEmpty()) { // TODO: Make a distinction between warnings and errors
            appliedChanges.addAll(changes);


        } else {
            restore();
        }
        return problems;
    }

    public List<Problem> load(List<Change<?>> changes) {
        loading = true;
        try {
            return apply(changes);
        } finally {
            loading = false;
        }
    }

    public boolean isLoading() {
        return loading;
    }

    private void restore() {
        nodes.clear();
        projects.clear();

        List<Change<?>> changes = new ArrayList<>(appliedChanges);
        appliedChanges.clear();
        apply(changes);
    }

    public <T extends Node> T node(String id) {
        return id == null ? null : (T) nodes.get(id);
    }

    public void addNode(Node<?> node) {
        nodes.put(node.id(), node);
        if (node instanceof ProjectNode) {
            projects.add((ProjectNode) node);
        }
        node.children().forEach(this::addNode);
    }

    public boolean isBatchUpdate() {
        return batchUpdate;
    }
}

File: src/main/kotlin/ludwig/workspace/Workspace.kt
package ludwig.workspace

import ludwig.changes.*
import ludwig.interpreter.Builtins
import ludwig.model.*
import ludwig.runtime.StdLib
import ludwig.script.Parser
import java.io.InputStreamReader
import java.util.*
import java.util.function.Consumer

class Workspace {

    private val nodes = HashMap<String, Node>()
    private val appliedChanges = mutableListOf<Change>()
    val projects = mutableListOf<ProjectNode>()
    private val changeListeners = mutableListOf<(Array<out Change>) -> Unit>()
    private val builtins = ProjectNode()
    var isLoading: Boolean = false
        private set

    fun init() {
        builtins.name = "Runtime"
        builtins.readonly = true
        builtins.id = "Runtime"
        builtins.add(Builtins.of(StdLib))

        addNode(builtins)

        try {
            InputStreamReader(Workspace::class.java.getResourceAsStream("/system.lw")).use { reader -> Parser.parse(reader, this, builtins) }
            InputStreamReader(Workspace::class.java.getResourceAsStream("/system-tests.lw")).use { reader -> Parser.parse(reader, this, builtins) }
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    fun changeListeners(): MutableList<(Array<out Change>) -> Unit> {
        return changeListeners
    }

    private fun place(node: Node, insert: Insert): Problem? {
        addNode(node)
        val parent = node(insert.parent)
        node.parent = parent

        if (parent != null) {
            val prev = node(insert.prev)
            val next = node(insert.next)

            if (!parent.isOrdered) {
                parent.add(node)
            } else {
                val items = parent

                if (next == null) {
                    if (!items.isEmpty() && items[items.size - 1] === prev || items.isEmpty() && prev == null) {
                        parent.add(node)
                    }
                } else if (prev == null) {
                    if (!items.isEmpty() && items[0] === next) {
                        items.add(0, node)
                    }
                } else {
                    val prevIndex = items.indexOf(prev)
                    val nextIndex = items.indexOf(next)

                    if (nextIndex == prevIndex + 1) {
                        items.add(nextIndex, next)
                    }
                }
            }
        }
        return null
    }

    fun apply(vararg changes: Change): List<Problem> {
        val problems = mutableListOf<Problem>()
        for (change in changes) {
            when (change) {
                is Create -> {
                    val node = Class.forName(Node::class.java.`package`.name + "." + change.nodeType).newInstance() as Node
                    node.id = change.changeId
                    place(node, change)
                }
                is Delete -> node(change.nodeId)!!.delete()
                is Comment -> node(change.nodeId)!!.comment = change.comment
                is Rename -> (node(change.nodeId) as NamedNode).name = change.name
                is Value -> {
                    val node = node(change.nodeId)
                    when (node) {
                        is LiteralNode -> node.text = change.value
                        is ReferenceNode -> node.ref = node(change.value)!!
                    }
                }
            }
        }

        changeListeners.forEach({ it(changes) })

        if (problems.isEmpty()) { // TODO: Make a distinction between warnings and errors
            appliedChanges.addAll(changes)


        } else {
            restore()
        }
        return problems
    }

    fun load(changes: Array<out Change>): List<Problem> {
        isLoading = true
        try {
            return apply(*changes)
        } finally {
            isLoading = false
        }
    }

    private fun restore() {
        nodes.clear()
        projects.clear()

        val changes = ArrayList(appliedChanges)
        appliedChanges.clear()
        apply(*changes.toTypedArray())
    }

    fun node(id: String?): Node? {
        return if (id == null) null else nodes[id]
    }

    fun addNode(node: Node) {
        nodes.put(node.id, node)
        if (node is ProjectNode) {
            projects.add(node)
        }
        node.forEach(Consumer { this.addNode(it) })
    }
}

fun changeListener(consumer: (Change) -> Unit): (Array<out Change>) -> Unit = { changes -> changes.forEach(consumer) }